### PR TITLE
[multibody] Parse screw joints from URDF

### LINKED ISSUE #17765: Add SDFormat and URDF parsing for screw joint
See https://github.com/RobotLocomotion/drake/issues/12404#issuecomment-1111679692

- [ ] URDF - will need custom joint https://github.com/RobotLocomotion/drake/pull/18191
- [x] SDFormat #18127

### PR DESCRIPTION
**Pull request by scpeters**
Closes #17765.

This adds support for parsing `screw` joints from URDF using a custom `<drake:joint/>` element. The screw thread pitch is specified with a `value` attribute in a custom `<drake:screw_thread_pitch/>` element. To maintain parity with the SDFormat parsing test, a continuous joint is added to the URDF parsing test to match the changes to the SDFormat test from #17956.

- - -
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18191)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
@amcastro-tri and @EricCousineau-TRI you reviewed #17956; would either of you like to sign up for this review likewise?

**Review by EricCousineau-TRI — COMMENTED** [review]
:lgtm_strong: feature, thanks! Just a nit on consistency of error message
+@jwnimmer-tri for platform review Tues, please!

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18191#-:-NFkyWN09T56-QRZrLOB:b-igfvkw)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @jwnimmer-tri and @scpeters)

___
*[`multibody/parsing/detail_urdf_parser.cc` line 572 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFkxqvYERdFxfQmhfQl:-NFkxqvYERdFxfQmhfQm:b7p3w6n) ([raw file](https://github.com/robotlocomotion/drake/blob/876a0ef7eb524d1007561c6f4df6f202ce9a74a7/multibody/parsing/detail_urdf_parser.cc#L572)):*
> ```cpp
>         Error(*screw_thread_pitch_node,
>               fmt::format("Joint {}'s drake:screw_thread_pitch does not have"
>                           " a \"value\" attribute.", name));
> ```

nit These error messages do not appear consistent with those above.

I see elements denoted as `<element>`, and attributes denoted as `'attribute'`.

**Review by EricCousineau-TRI — COMMENTED** [review]
+@ericcousineau-tri 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18191#-:-NFkylYD5xuWiJ0jygSt:b-hywmtj)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @jwnimmer-tri and @scpeters)

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm: platform.

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18191#-:-NFnmpGD5o0UOnWlemRb:b-vtj4vo)* status: 4 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @scpeters)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFnk30Z4GzegSQFW5XE:-NFnk30_2CAra_4bYztS:bh9r1wg) (no related file):*
The parser's [documentation](https://drake.mit.edu/doxygen_cxx/group__multibody__parsing.html) needs updating to reflect the new elements/attributes in this PR.

FYI See https://drake.mit.edu/documentation_instructions.html for local preview.

___
___
*[`multibody/parsing/detail_urdf_parser.cc` line 563 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFnjXc-4lqQW__8Nvie:-NFnjXc03ORvZ0GX_RUs:bedsdfc) ([raw file](https://github.com/robotlocomotion/drake/blob/876a0ef7eb524d1007561c6f4df6f202ce9a74a7/multibody/parsing/detail_urdf_parser.cc#L563)):*
> ```cpp
>     throw_on_custom_joint(true);
>     ParseJointDynamics(node, &damping);
>     // Parse screw thread pitch
> ```

nit Missing punctuation on a complete sentence.



_Suggestion:_
```cpp
// Parse screw thread pitch.
```
___
*[`multibody/parsing/detail_urdf_parser.cc` line 564 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFnkT7D9ZpePPXnm60N:-NFnkT7D9ZpePPXnm60O:b-ikm29a) ([raw file](https://github.com/robotlocomotion/drake/blob/876a0ef7eb524d1007561c6f4df6f202ce9a74a7/multibody/parsing/detail_urdf_parser.cc#L564)):*
> ```cpp
>     ParseJointDynamics(node, &damping);
>     // Parse screw thread pitch
>     double screw_thread_pitch;
> ```

Looking around within this parsing function, it seems to me like the local style is that long-winded branchy code like this new thread-pitch parsing should be delegated to helper functions, so that the main if-else chain in this function is straightforward and easy to read.

Also because I think it will help in that refactor -- in case of bad input, I think it's fine for the pitch-parsing helper to fire off an Error and then return some nominal value like 0.001 or 0.0 that will placate the ScrewJoint constructor so that we can still add the Joint to the plant.  That will likely make the control flow easier, and having the Joint object with the correct name in the tree might reduce additional spurious error messages later on.

**Review by scpeters — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18191)* status: 4 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @EricCousineau-TRI and @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFnk30Z4GzegSQFW5XE:-NFrR6AK1v8XIwEpe8FH:b-94ddnn) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The parser's [documentation](https://drake.mit.edu/doxygen_cxx/group__multibody__parsing.html) needs updating to reflect the new elements/attributes in this PR.

FYI See https://drake.mit.edu/documentation_instructions.html for local preview.
</blockquote></details>

Ok, I've tried updating it.

___
___
*[`multibody/parsing/detail_urdf_parser.cc` line 563 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFnjXc-4lqQW__8Nvie:-NFrRK4J7mN0C0fkg-6y:b-fyi9zb) ([raw file](https://github.com/robotlocomotion/drake/blob/876a0ef7eb524d1007561c6f4df6f202ce9a74a7/multibody/parsing/detail_urdf_parser.cc#L563)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit Missing punctuation on a complete sentence.
</blockquote></details>

I ended up removing this comment
___
*[`multibody/parsing/detail_urdf_parser.cc` line 564 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFnkT7D9ZpePPXnm60N:-NFrRASBCjPQfzqfSF-n:bw9hcx2) ([raw file](https://github.com/robotlocomotion/drake/blob/876a0ef7eb524d1007561c6f4df6f202ce9a74a7/multibody/parsing/detail_urdf_parser.cc#L564)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Looking around within this parsing function, it seems to me like the local style is that long-winded branchy code like this new thread-pitch parsing should be delegated to helper functions, so that the main if-else chain in this function is straightforward and easy to read.

Also because I think it will help in that refactor -- in case of bad input, I think it's fine for the pitch-parsing helper to fire off an Error and then return some nominal value like 0.001 or 0.0 that will placate the ScrewJoint constructor so that we can still add the Joint to the plant.  That will likely make the control flow easier, and having the Joint object with the correct name in the tree might reduce additional spurious error messages later on.
</blockquote></details>

Good call; I moved the logic to a ParseScrewJointThreadPitch method
___
*[`multibody/parsing/detail_urdf_parser.cc` line 572 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFkxqvYERdFxfQmhfQl:-NFrRFtS5mB92o_LBQ28:ba89kwa) ([raw file](https://github.com/robotlocomotion/drake/blob/876a0ef7eb524d1007561c6f4df6f202ce9a74a7/multibody/parsing/detail_urdf_parser.cc#L572)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

nit These error messages do not appear consistent with those above.

I see elements denoted as `<element>`, and attributes denoted as `'attribute'`.
</blockquote></details>

Ok, I updated to follow this style

**Review by jwnimmer-tri — COMMENTED** [review]
+@rpoyner-tri for `parsing_doxygen.h` SME feature review, please.

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18191#-:-NFsbLt45MsWRPbORVsN:b-azwwxi)* status: 2 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @EricCousineau-TRI, @rpoyner-tri, and @scpeters)

___
*[`multibody/parsing/detail_urdf_parser.cc` line 420 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFsArVj2A1c9HX8DZBU:-NFsa8qjBC6QfNGlviA9:bxpcf8z) ([raw file](https://github.com/robotlocomotion/drake/blob/36d112627aa5d92e1e9bd808c13ec04a2d94f6a0/multibody/parsing/detail_urdf_parser.cc#L420)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

I'm not sure if I get this comment.

How is there a default value if this attribute is required?
Also, `0.0 != 1`?
</blockquote></details>

Maybe a clearer comment would be "Always set a value for the output-only argument, even if parsing fails."
___
*[`multibody/parsing/parsing_doxygen.h` line 714 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFsb0xF7zNFEg6C8JyW:-NFsb0xF7zNFEg6C8JyX:bn2bgdu) ([raw file](https://github.com/robotlocomotion/drake/blob/36d112627aa5d92e1e9bd808c13ec04a2d94f6a0/multibody/parsing/parsing_doxygen.h#L714)):*
> ```objc
> @ref reflected_inertia "Reflected Inertia"
> 
> */
> ```

nit I think we need two new `@subsection`s here documenting the new attribute, so that the `@ref`s above actually go somewhere.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18191)* status: 2 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @EricCousineau-TRI)

**Review by scpeters — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18191)* status: 2 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @EricCousineau-TRI, @jwnimmer-tri, and @rpoyner-tri)

___
*[`multibody/parsing/detail_urdf_parser.cc` line 420 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFsArVj2A1c9HX8DZBU:-NFvjTf23-668mXAAfHO:b9lmtv5) ([raw file](https://github.com/robotlocomotion/drake/blob/36d112627aa5d92e1e9bd808c13ec04a2d94f6a0/multibody/parsing/detail_urdf_parser.cc#L420)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Maybe a clearer comment would be "Always set a value for the output-only argument, even if parsing fails."
</blockquote></details>

Ok, I've applied Jeremy's suggestion.
___
*[`multibody/parsing/parsing_doxygen.h` line 714 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18191#-NFsb0xF7zNFEg6C8JyW:-NFvjP4U1Svilk_3p4SV:b-cibw80) ([raw file](https://github.com/robotlocomotion/drake/blob/36d112627aa5d92e1e9bd808c13ec04a2d94f6a0/multibody/parsing/parsing_doxygen.h#L714)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

Actually, a `@subsection tag_drake_screw_thread_pitch drake:screw_thread_pitch` and a `@subsubsection tag_drake_screw_thread_pitch_semantics Semantics`.
</blockquote></details>

Ok, I've added the requested subsections and fixed a few incorrect XPath expressions that I noticed elsewhere in the file

**Review by EricCousineau-TRI — COMMENTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18191)* status: 1 unresolved discussion, LGTM missing from assignee rpoyner-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @rpoyner-tri)

**Review by rpoyner-tri — COMMENTED** [review]
:lgtm:

> Reviewed 3 of 5 files at r1, 1 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18191#-:-NFxr6P3AJx8dvkR3KB7:bnfp4nl)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @scpeters)

File: multibody/parsing/parsing_doxygen.h [hunk 1]
```diff
@@ -708,4 +711,19 @@ object. Units are kg⋅m² for revolute joints, and kg for prismatic joints.
 @see drake::multibody::JointActuator, @ref tag_drake_gear_ratio,
 @ref reflected_inertia "Reflected Inertia"
 
+@subsection tag_drake_screw_thread_pitch drake:screw_thread_pitch
+
+- SDFormat path: `//model/joint/drake:screw_thread_pitch`
```
Comment (EricCousineau-TRI) [inline multibody/parsing/parsing_doxygen.h:716]: I'll file chaser PR to fix this to `//model/joint/screw_thread_pitch` given that it's part of SDFormat proper

**Pull request merged**